lib/api/units/validateBody.js

```diff
@@ -0,0 +1,233 @@
+const jph = require('json-parse-helpfulerror');
+const mediaTyper = require('media-typer');
+const contentTypeUtils = require('content-type');
+
+const { TextDiff } = require('../../../lib/validators/text-diff');
+const { JsonExample } = require('../../../lib/validators/json-example');
+const { JsonSchema } = require('../../../lib/validators/json-schema');
+
+function isPlainText(mediaType) {
+  return mediaType.type === 'text' && mediaType.subtype === 'plain';
+}
+
+function isJson(mediaType) {
+  return mediaType.type === 'application' && mediaType.subtype === 'json';
+}
+
+function isJsonWeak(mediaType) {
+  return isJson(mediaType) || mediaType.suffix === 'json';
+}
+
+/**
+ * Parses a given content-type header into media type.
+ * @param {string} contentType
+ * @returns {[Error, MediaType]}
+ */
+function parseContentType(contentType) {
+  try {
+    const { type } = contentTypeUtils.parse(`${contentType}`);
+    const mediaType = mediaTyper.parse(type);
+    return [null, mediaType];
+  } catch (error) {
+    return [error, null];
+  }
+}
+
+/**
+ * Determines if a given 'Content-Type' header contains JSON.
+ * @param {string} contentType
+ * @returns {boolean}
+ */
+function isJsonContentType(contentType) {
+  // No need to explicitly check for "contentType" existence.
+  // In case given content type is invalid, "parseContentType"
+  // will return an error, which would be coerced to "false".
+  const [error, mediaType] = parseContentType(contentType);
+
+  // Silence an error on purporse because contentType
+  // may contain any kind of rubbish.
+  return error ? false : isJsonWeak(mediaType);
+}
+
+/**
+ * Returns a tuple of error and body media type based
+ * on the given body and normalized headers.
+ * @param {string} body
+ * @param {Object} headers
+ * @returns {[error, bodyType]}
+ */
+function getBodyType(body, headers) {
+  const contentType = headers && headers['content-type'];
+  const hasJsonContentType = isJsonContentType(contentType);
+
+  try {
+    jph.parse(body);
+    const [mediaTypeError, bodyMediaType] = parseContentType(
+      hasJsonContentType ? contentType : 'application/json'
+    );
+
+    if (mediaTypeError) {
+      throw new Error(mediaTypeError);
+    }
```

Comment:
I think this `throw` here is not needed at all. At least, it doesn't align with the nature of the thrown error. Catch block below handled scenarios when `body` is not a parsable JSON. However, if `parseContentType` fails, this already means that the body has been parsed successfully. `mediaTypeError` should be handled separately (if it's even needed).

This is one more argument to not mix errors context, since in your example both failing `jph.parse` and `parseContentType` are handled with a single `catch` block, but bear different nature of errors. To provide respective handling for those errors one would introduce nesting of try/catch, and I'm somewhat against that (see the points above).

Reply:
Removed that throw block https://github.com/apiaryio/gavel.js/pull/155/commits/9c50d2f1216b5e1d47ed7d1eaaab28bf8db00fca.